aspnetcore/performance/caching/memory/sample/WebCache/Controllers/HomeController.cs

```diff
@@ -0,0 +1,213 @@
+            .RegisterPostEvictionCallback(postEvictDelegate, state: this);
+    }
+
+    // Show key value and why key was evicted.
+    private static void AfterEvicted(object key, object value,
+        EvictionReason reason, object state)
+    {
+        var em = $"key: {key}, Value: {value}, Reason: {reason}";
+        ((HomeController)state)._memoryCache.Set<string>(CacheKey.EvictMsg1, em);
+    }
+
+    public IActionResult CheckEvictionTime()
+    {
+        ViewData["Message"] = _memoryCache.Get<string>(CacheKey.EvictMsg1);
+
+        return View(_memoryCache.Get<DateTime>(CacheKey.MS));
+    }
+    #endregion
+
+
+    #region snippet_ed
+    public IActionResult EvictDependency()
+    {
+        // Clear out eviction message, and key from previous run (if any).
+        _memoryCache.Remove(CacheKey.EvictMsg2);
+        CancellationTokenSource cts2 = new CancellationTokenSource();
+        _memoryCache.Set<CancellationTokenSource>(CacheKey.CancelTokenSource2, cts2);
+
+        using (var entry = _memoryCache.CreateEntry(CacheKey.MS2))
+        {
+            // expire this entry if the dependant entry expires.
+            entry.Value = DateTime.Now.TimeOfDay.Milliseconds.ToString();
+            entry.RegisterPostEvictionCallback(AfterEvicted2, this);
+
+            _memoryCache.Set(CacheKey.MS3,
+                DateTime.Now.AddMilliseconds(4).TimeOfDay.Milliseconds.ToString(),
+                new CancellationChangeToken(cts2.Token));
+        }
+
+        return RedirectToAction("CheckEvictDependency");
+    }
+
+    public IActionResult CheckEvictDependency(int? id)
+    {
+        ViewData["CachedMS2"] = _memoryCache.Get<string>(CacheKey.MS2);
+        ViewData["CachedMS3"] = _memoryCache.Get<string>(CacheKey.MS3);
+        ViewData["MessageCED"] = _memoryCache.Get<string>(CacheKey.EvictMsg2);
+
+        if (id > 0)
+        {
+            CancellationTokenSource cts2 =
+                _memoryCache.Get<CancellationTokenSource>(CacheKey.CancelTokenSource2);
+            cts2.Cancel();
+        }
+
+        return View();
+    }
+    #endregion
+
+    private static void AfterEvicted2(object key, object value,
+                                      EvictionReason reason, object state)
+    {
+        var em = $"key: {key}, Value: {value}, Reason: {reason}";
+        ((HomeController)state)._memoryCache.Set<string>(CacheKey.EvictMsg2, em);
+    }
+     
+
+    public IActionResult CheckCancel(int? id = 0)
+    {
+        if (id > 0)
+        {
+            CancellationTokenSource cts =
+               _memoryCache.Get<CancellationTokenSource>(CacheKey.CancelTokenSource);
+            cts.CancelAfter(100);
+            // Cancel immediately with cts.Cancel();
+        }
+
+        ViewData["CachedTime"] = _memoryCache.Get<string>(CacheKey.Ticks);
+        ViewData["Message"] =  _memoryCache.Get<string>(CacheKey.CancelMsg); ;
+
+        return View();
+    }
+    public IActionResult CancelTest()
```

Comment:
A few of these actions don't seem to contribute to the sample much. They are not referenced in the doc and I don't see what additional usage they illustrate.